module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,207 @@
+package indexer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/migrations"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// ExecutionState indexes the execution state.
+type ExecutionState struct {
+	registers  storage.RegisterIndex
+	headers    storage.Headers
+	events     storage.Events
+	indexRange *SequentialIndexRange
+	log        zerolog.Logger
+}
+
+// New execution state indexer used to ingest block execution data and index it by height.
+// The passed RegisterIndex storage must be populated to include the first and last height otherwise the indexer
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*ExecutionState, error) {
+	// we must have the first height set in the storage otherwise error out
+	first, err := registers.FirstHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing first height")
+	}
+
+	// we must have the latest height set in the storage otherwise error out
+	last, err := registers.LatestHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing latest height")
+	}
+
+	indexRange, err := NewSequentialIndexRange(first, last)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug().
+		Uint64("first_height", first).
+		Uint64("latest_height", last).
+		Msgf("initialized indexer range")
+
+	return &ExecutionState{
+		registers:  registers,
+		headers:    headers,
+		events:     events,
+		indexRange: indexRange,
+		log:        log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	_, err := i.indexRange.Contained(height)
```

Comment:
I've noticed we don't cross check the value of `pebble.LatestHeight()` which can provide this check implicitly.

Reply:
Not sure I understand what you mean?